lib/recursive-methods.rb

```diff
@@ -1,49 +1,94 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - n recursive calls ont he stack
+# Space complexity: O(n) - n recursive calls on the stack
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  return 1 if n == 0
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2) - n recursive calls, and slicing the string in each call takes n time/space
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s
+    end
+
+    return s[-1] + reverse(s[1..-2]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2) - because n recursive calls, inside which we are slicing smaller...
+# Space complexity: O(n)??? because it is just saving over the same variable?
+
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length <= 1
+
+  s.replace(s[-1] + reverse_inplace(s[1..-2]) + s[0])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - n recursive calls
+# Space complexity: O(n) - n recursive calls, all else constant
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  return 0 if n == 0
+  return 2 if n == 1
+
+  return bunny(n - 1) + 2
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - really n/2 recursive calls, and the string returned is smaller each time until it returns false or returns true at the middle of the string length
+# Space complexity: O(n)? n/2 recursive calls, also slicing the string each time...
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    if s == "" #this also works: s.length == 0
+      return true
+    elsif s[0] == "(" && s[-1] == ")"
+      return nested(s[1..-2])
+    else
+      return false
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - call stack is length n, but rest of the code is constant
+# Space complexity: O(n)??? the call stacK?  but also array.drop(1), is basically returning the array each time (albeit smaller -1) so maybe this is O(n^2) for space?
 def search(array, value)
```

Comment:
👍 However this will remove elements from the array which means it mutates the array with `.drop` which can be problematic if this method was used for other purposes.